src/sys/windows/selector.rs

```diff
@@ -221,14 +286,36 @@ impl Registration {
         Ok(())
     }
 
-    pub fn deregister(&mut self, _selector: &mut Selector) -> io::Result<()> {
-        // TODO: assert that self.selector == selector?
-        if self.selector.is_none() {
-            Err(super::bad_state())
-        } else {
-            self.selector = None;
-            Ok(())
+    pub fn deregister(&mut self) {
+        trace!("deregister; token={:?}", self.token);
+
+        if let Some(entry) = self.level.take() {
+            let selector = self.selector.as_ref()
+                .expect("expected selector");
+
+            selector.level_triggered.lock().unwrap().remove(entry);
+        }
+
+        self.selector = None;
+    }
+
+    pub fn checked_deregister(&mut self, selector: &Selector) -> io::Result<()> {
+        match self.selector {
+            Some(ref s) => {
+                let inner1: &SelectorInner = &*selector.inner;
+                let inner2: &SelectorInner = &*s;
+
+                if inner1 as *const SelectorInner != inner2 as *const SelectorInner {
+                    return Err(other("socket registered with other selector"));
```

Comment:
May be good to add a test for this as well? Although that can also probably be tracked on the issue as well.